### PR TITLE
Add support for fetching from JSON columns

### LINKED ISSUE #129: 💡 Support for Querying JSON Columns in Databases for better DX
### I have an idea!

# Feature Request:

I am requesting the inclusion of enhanced methods to query JSON column types directly via cycle/orm. This feature not only aligns with the capabilities offered by frameworks like Laravel, but also significantly improves the overall Developer Experience (DX).

## Current Behavior:

Currently, querying JSON in cycle/orm involves more complex, less intuitive code:

```php
...

use Cycle\Database\Injection\Parameter;

...

$exists = $webhookCallsRepository
    ->select()
    ->where(['name' => 'stripe'])
    ->andWhere("JSON_EXTRACT(payload, '$.id')", '=', new Parameter(['payloadId' => $request->get('id')]))
    ->count();
```

Full class located here: https://github.com/wayofdev/laravel-stripe-webhooks/blob/master/src/Profile/StripeWebhookProfile.php

**Desired Behavior:**

```php
$users = $repository->select()
                ->whereJson('preferences->dining->meal', 'salad')
                ->fetch();

$users = $repository->select()
                ->whereJsonContains('options->languages', 'en')
                ->fetch();

$users = $repository->select()
                ->whereJsonLength('options->languages', '>', 1)
                ->fetch();
```

## Proposal:

I've checked internal code of cycle/orm and found that first thing to modify would be WhereTrait.php, where additional methods could be implemented:

```php
<?php

trait WhereTrait
{
    // ... (other existing methods) ...
    
    public function whereJson(string $column, $value): self
    {
        $this->registerToken(
            'AND',
            [$column, '->', $value],
            $this->whereTokens,
            $this->whereWrapper()
        );

        return $this;
    }

    public function whereJsonContains(string $column, $value): self
    {
        $this->registerToken(
            'AND',
            [$column, '->', $value],
            $this->whereTokens,
            $this->whereWrapper()
        );

        return $this;
    }

    public function whereJsonLength(string $column, $operator, $value = null): self
    {
        if ($value === null) {
            $value = $operator;
            $operator = '=';
        }

        $this->registerToken(
            'AND',
            [$column, 'json_length', $operator, $value],
            $this->whereTokens,
            $this->whereWrapper()
        );

        return $this;
    }
}
```

## Benefits:

**Improved Developer Experience (DX):** This change would significantly reduce the complexity and verbosity of JSON-related queries, leading to faster and more intuitive coding.

**Reduced Errors:** Leveraging built-in methods reduces the likelihood of errors from manually written SQL statements.

**Alignment with Popular Frameworks:** It brings cycle/orm on par with other frameworks, making it a more attractive option for developers familiar with those patterns.
I believe this feature, coupled with other necessary changes, would be a valuable addition to cycle/orm. I look forward to feedback and potential inclusion in a future release.

### PR DESCRIPTION
**Pull request by msmakouz**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ❌
| New feature?  | ✔️
| Issues        | #129

This PR introduces a set of new query methods. These methods are designed to simplify working with **JSON** columns in the database, making it easier to filter and manipulate data stored in JSON format.

### 1) WhereJson

- `whereJson(string $path, mixed $value): static` - allow to select records where the JSON value at the specified path is equal to the passed value.
- `orWhereJson(string $path, mixed $value): static` - similar to the `whereJson` method, but uses the **OR** operator to combine with other conditions.

```php
$users = $database
    ->select()
    ->from('users')
    ->whereJson('options->notifications->type', 'sms')
    ->fetchAll();

$users = $database
    ->select()
    ->from('users')
    ->whereJson('options->notifications->type', 'email')
    ->orWhereJson('options->notifications->type', 'sms')
    ->fetchAll();
```

### 2) WhereJsonContains

- `whereJsonContains(string $path, mixed $value): static` - allow to select records where the JSON array contains the passed value. The passed value can be an array, which will be encoded into JSON.
- `orWhereJsonContains(string $path, mixed $value): static` - similar to the `whereJsonContains` method, but uses the **OR** operator to combine with other conditions.

```php
$users = $database
    ->select()
    ->from('users')
    ->whereJsonContains('options->languages', 'en')
    ->fetchAll();

$users = $database
    ->select()
    ->from('users')
    ->where('status', 'active')
    ->orWhereJsonContains('options->languages', 'fr')
    ->fetchAll();
```

### 3) WhereJsonDoesntContain

- `whereJsonDoesntContain(string $path, mixed $value): static` - allow to select records where the JSON array does **NOT** contain the passed value. The passed value can be an array, which will be encoded into JSON.
- `orWhereJsonDoesntContain(string $path, mixed $value): static` - similar to the `whereJsonDoesntContain` method, but uses the **OR** operator to combine with other conditions.

```php
$users = $database
    ->select()
    ->from('users')
    ->whereJsonDoesntContain('options->languages', 'de')
    ->fetchAll();

$users = $database
    ->select()
    ->from('users')
    ->whereJsonDoesntContain('options->languages', 'fr')
    ->orWhereJsonDoesntContain('options->languages', 'de')
    ->fetchAll();
```

### 4) WhereJsonContainsKey

- `whereJsonContainsKey(string $path): static` - allows to select records where the JSON contains the desired path.
- `orWhereJsonContainsKey(string $path): static` - method is used to combine with other conditions using the **OR** operator.

```php
$users = $database
    ->select()
    ->from('users')
    ->whereJsonContainsKey('options->languages->de')
    ->fetchAll();

$users = $database
    ->select()
    ->from('users')
    ->orWhereJsonContainsKey('options->languages->fr')
    ->orWhereJsonContainsKey('options->languages->de')
    ->fetchAll();
```

### 5) WhereJsonDoesntContainKey

- `whereJsonDoesntContainKey(string $path): static` - allows to select records where the JSON does **NOT** contain the desired path.
- `orWhereJsonDoesntContainKey(string $path): static` - method is used to combine with other conditions using the **OR** operator.

```php
$users = $database
    ->select()
    ->from('users')
    ->whereJsonDoesntContainKey('options->languages->de')
    ->fetchAll();

$users = $database
    ->select()
    ->from('users')
    ->orWhereJsonDoesntContainKey('options->languages->fr')
    ->orWhereJsonDoesntContainKey('options->languages->de')
    ->fetchAll();
```

### 6) WhereJsonLength

- `whereJsonLength(string $path, int $length, string $operator = '='): static` - allows to select JSON arrays by their length.
- `orWhereJsonLength(string $path, int $length, string $operator = '='): static` - method is used to combine with other conditions using the **OR** operator.

```php
$users = $database
    ->select()
    ->from('users')
    ->whereJsonLength('journal->errors', 0)
    ->fetchAll();

$users = $database
    ->select()
    ->from('users')
    ->whereJsonLength('journal->errors', 1, '>=')
    ->fetchAll();
```

**All of these conditions can also be used in delete and update queries.**

### CODE REVIEW

File: src/Injection/JsonExpression.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            }
+
+            return $parts[0];
+        }
+
+        return '"' . $segment . '"';
+    }
+
+    /**
+     * @param non-empty-string $attribute
```
Comment (roxblnfk) [inline src/Injection/JsonExpression.php]: Please add more comments about functions. What kind of strings it parses and what result returns. I can't find unit tests with this class

**Review by roxblnfk — CHANGES_REQUESTED** [review]

File: src/Query/Traits/WhereJsonTrait.php [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+declare(strict_types=1);
+
+namespace Cycle\Database\Query\Traits;
+
+use Cycle\Database\Exception\BuilderException;
+
+trait WhereJsonTrait
+{
+    /**
+     * @param non-empty-string $column
```
Comment (roxblnfk) [inline src/Query/Traits/WhereJsonTrait.php]: Need to add comments to all these methods

Comment (msmakouz) [reply src/Query/Traits/WhereJsonTrait.php]: Added

**Review by roxblnfk — APPROVED** [review]
LGTM

**Pull request merged**